### ISSUE TITLE
optimise API timeouts

### ISSUE THREAD

**Issue by stefangotz** [echronos/echronos#7]
# Motivation

Some RTOS components implement APIs that include a timeout parameter.
For example, the `message_queue_get_timeout()` API does not block indefinitely until a message is available, but only until the specified timeout expires.
The mutex and semaphore components offer similar APIs.
These APIs are implemented by waiting on a signal that is raised either by another task to indicate an event (such as a new message being available in a queue) or by a timer to indicate that the specified timeout expired.

When a task A is signaled by another task B, the condition that task A is waiting for may not be true despite the signal (for example, another task C might have been signaled, too, and might have consumed the message before task A had a chance to).
In this case, task A restarts its timer and blocks again.
Restarting the timer is not strictly necessary because it has not yet expired yet and is still active with the original, correct timeout value.
Therefore, it would actually still expire at the intended point in time.

There is an optimization opportunity in making the API implementations activate the timeout timer only once and rely on it to expire without restarting it.
This could potentially save both execution overhead and code space.

# Goals

- explore whether the optimization outlined above would yield tangible benefits
- if so, implement this optimization
- create regression tests for the timeout APIs of the mutex, semaphore, and message queue components where necessary

# Test Plan

- review code modifications and regression tests
- all regression tests must pass
